File: client/Sources/Features/Git/GitStatusModel.swift
// client/Sources/Features/Git/GitStatusModel.swift
import Foundation
import Observation
import MaverickProtocol

/// Drives the git diff tab. Holds the current GitStatus + a map of cached
/// diffs per file. Diffs are fetched on demand (when the user taps a file
/// in the list).
@Observable
final class GitStatusModel {
    enum State: Equatable { case idle, loading, loaded, error(String) }

    private(set) var path: String = ""
    private(set) var state: State = .idle
    private(set) var status: GitStatus = .notARepo

    /// Diff text per file (UI side cache, keyed by "<staged?>:<path>" so
    /// staged vs working-tree diffs don't collide).
    private(set) var diffs: [String: DiffResult] = [:]
    /// Files currently being fetched, for spinner display.
    private(set) var pendingDiffs: Set<String> = []

    struct DiffResult: Equatable {
        let text: String
        let truncated: Bool
    }

    private var pendingStatusRequestId: UUID?
    private var pendingDiffRequests: [UUID: String] = [:]   // requestId -> diffKey

    static func key(file: String, staged: Bool) -> String { (staged ? "S:" : "U:") + file }

    func refresh(path: String, connection: ConnectionManager) {
        let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        let req = UUID()
        pendingStatusRequestId = req
        state = .loading
        self.path = trimmed
        connection.send(.gitStatus(requestId: req, path: trimmed))
    }

    func fetchDiff(file: String, staged: Bool, connection: ConnectionManager) {
        let key = Self.key(file: file, staged: staged)
        // Skip if already cached or in flight.
        if diffs[key] != nil { return }
        if pendingDiffs.contains(key) { return }
        let req = UUID()
        pendingDiffRequests[req] = key
        pendingDiffs.insert(key)
        connection.send(.gitDiff(requestId: req, path: path, file: file, staged: staged))
    }

    func handle(_ message: ServerMessage) {
        switch message {
        case .gitStatusResult(let reqId, let s):
            guard reqId == pendingStatusRequestId else { return }
            status = s
            state = .loaded
            pendingStatusRequestId = nil
        case .gitStatusFailed(let reqId, let msg):
            guard reqId == pendingStatusRequestId else { return }
            state = .error(msg)
            pendingStatusRequestId = nil
        case .gitDiffResult(let reqId, _, let diff, let truncated):
            guard let key = pendingDiffRequests.removeValue(forKey: reqId) else { return }
            pendingDiffs.remove(key)
            diffs[key] = DiffResult(text: diff, truncated: truncated)
        case .gitDiffFailed(let reqId, let msg):
            guard let key = pendingDiffRequests.removeValue(forKey: reqId) else { return }
            pendingDiffs.remove(key)
            diffs[key] = DiffResult(text: "[diff failed] \(msg)", truncated: false)
        default:
            break
        }
    }
}

File: mobile/src/stores/git-status-model.ts
/**
 * Port of the Swift `GitStatusModel`. Drives the git diff screen: holds the
 * current `GitStatus` plus a per-file diff cache keyed by `<staged?>:<file>`,
 * fetching diffs on demand and correlating replies by requestId.
 */

import { MaverickClient } from '@/app/maverick-client';
import { Observable } from '@/lib/observable';
import { GitStatus, ServerMessage } from '@/protocol';

export type GitState =
  | { kind: 'idle' }
  | { kind: 'loading' }
  | { kind: 'loaded' }
  | { kind: 'error'; message: string };

export interface DiffResult {
  text: string;
  truncated: boolean;
}

const NOT_A_REPO: GitStatus = {
  isRepo: false,
  ahead: 0,
  behind: 0,
  files: [],
};

export function diffKey(file: string, staged: boolean): string {
  return (staged ? 'S:' : 'U:') + file;
}

export class GitStatusModel extends Observable {
  private repoPath = '';
  private currentState: GitState = { kind: 'idle' };
  private currentStatus: GitStatus = NOT_A_REPO;
  private readonly diffCache = new Map<string, DiffResult>();
  private readonly pendingDiffKeys = new Set<string>();

  private pendingStatusRequestId: string | null = null;
  private readonly pendingDiffRequests = new Map<string, string>();

  constructor(private readonly client: MaverickClient) {
    super();
  }

  get path(): string {
    return this.repoPath;
  }
  get state(): GitState {
    return this.currentState;
  }
  get status(): GitStatus {
    return this.currentStatus;
  }
  diff(file: string, staged: boolean): DiffResult | undefined {
    return this.diffCache.get(diffKey(file, staged));
  }
  isDiffPending(file: string, staged: boolean): boolean {
    return this.pendingDiffKeys.has(diffKey(file, staged));
  }

  refresh(path: string): void {
    const trimmed = path.trim();
    if (trimmed.length === 0) return;
    const req = this.client.requestId();
    this.pendingStatusRequestId = req;
    this.currentState = { kind: 'loading' };
    this.repoPath = trimmed;
    this.notify();
    this.client.gitStatus(req, trimmed);
  }

  fetchDiff(file: string, staged: boolean): void {
    const key = diffKey(file, staged);
    if (this.diffCache.has(key) || this.pendingDiffKeys.has(key)) return;
    const req = this.client.requestId();
    this.pendingDiffRequests.set(req, key);
    this.pendingDiffKeys.add(key);
    this.notify();
    this.client.gitDiff(req, this.repoPath, file, staged);
  }

  handle(message: ServerMessage): void {
    switch (message.type) {
      case 'git_status_result':
        if (message.requestId !== this.pendingStatusRequestId) return;
        this.currentStatus = message.status;
        this.currentState = { kind: 'loaded' };
        this.pendingStatusRequestId = null;
        this.notify();
        break;
      case 'git_status_failed':
        if (message.requestId !== this.pendingStatusRequestId) return;
        this.currentState = { kind: 'error', message: message.message };
        this.pendingStatusRequestId = null;
        this.notify();
        break;
      case 'git_diff_result': {
        const key = this.pendingDiffRequests.get(message.requestId);
        if (key === undefined) return;
        this.pendingDiffRequests.delete(message.requestId);
        this.pendingDiffKeys.delete(key);
        this.diffCache.set(key, {
          text: message.diff,
          truncated: message.truncated,
        });
        this.notify();
        break;
      }
      case 'git_diff_failed': {
        const key = this.pendingDiffRequests.get(message.requestId);
        if (key === undefined) return;
        this.pendingDiffRequests.delete(message.requestId);
        this.pendingDiffKeys.delete(key);
        this.diffCache.set(key, {
          text: `[diff failed] ${message.message}`,
          truncated: false,
        });
        this.notify();
        break;
      }
      default:
        break;
    }
  }
}
